lib/src/network/protocol/proxy_protocol/header.rs

```diff
@@ -130,36 +130,170 @@ pub struct HeaderV2 {
   ver_and_cmd: u8,     // protocol version and command
   family: u8,          // protocol family and address
   len: u16,            // number of following bytes part of the header
-  addr: AddrV2,
+  addr: ProxyAddr,
 }
 
 impl HeaderV2 {
-  fn new(ver_and_cmd: u8, family: u8, addr: AddrV2) -> Self {
+  pub fn new(addr_src: SocketAddr, addr_dst: SocketAddr) -> Self {
+    let addr = ProxyAddr::from(addr_src, addr_dst);
+
     HeaderV2 { 
       signature: [0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A],
-      ver_and_cmd,
-      family,
-      len: 0,
+      ver_and_cmd: 0x20,
+      family: get_family(&addr),
+      len: addr.len(),
       addr,
     }
   }
+
+  pub fn into_bytes(&self) -> Vec<u8> {
+    let mut header = Vec::new();
+    header.extend_from_slice(&self.signature);
+    header.push(self.ver_and_cmd);
+    header.push(self.family);
+    header.extend_from_slice(&u16_to_array_of_u8(self.len));
+    header.extend_from_slice(&self.addr.into_bytes());
+    header
+  }
 }
 
-pub enum AddrV2 {
-  // len = 12
-  TCP_UDP_IPV4 {
-    addr_src: SocketAddrV4,
-    addr_dst: SocketAddrV4,
+pub enum ProxyAddr {
+  Ipv4Addr {
+    src_addr: SocketAddrV4,
+    dst_addr: SocketAddrV4,
   },
-  // len = 36
-  TCP_UDP_IPV6 {
-    addr_src: SocketAddrV6,
-    addr_dst: SocketAddrV6,
+  Ipv6Addr {
+    src_addr: SocketAddrV6,
+    dst_addr: SocketAddrV6,
   },
-  // len = 216
-  AF_UNIX {
+  UnixAddr {
     src_addr: [u8; 108],
-    dest_addr: [u8; 108],
+    dst_addr: [u8; 108],
   },
-  AF_UNSPEC
+  AfUnspec,
 }
+
+impl ProxyAddr {
+
+  pub fn from(addr_src: SocketAddr, addr_dst: SocketAddr) -> Self {
+    match (addr_src, addr_dst) {
+      (SocketAddr::V4(addr_ipv4_src), SocketAddr::V4(addr_ipv4_dst)) => {
+        ProxyAddr::Ipv4Addr{
+          src_addr: addr_ipv4_src,
+          dst_addr: addr_ipv4_dst,
+        }
+      },
+      (SocketAddr::V6(addr_ipv6_src), SocketAddr::V6(addr_ipv6_dst)) => {
+        ProxyAddr::Ipv6Addr{
+          src_addr: addr_ipv6_src,
+          dst_addr: addr_ipv6_dst,
+        }
+      },
+      _ => ProxyAddr::AfUnspec,
+    }
+  }
+
+  fn len(&self) -> u16 {
+    match *self {
+      ProxyAddr::Ipv4Addr{ src_addr: _, dst_addr: _ } => 12,
+      ProxyAddr::Ipv6Addr{ src_addr: _, dst_addr: _ } => 36,
+      ProxyAddr::UnixAddr{ src_addr: _, dst_addr: _ } => 216,
+      ProxyAddr::AfUnspec => 0,
+    }
+  }
+
+  fn into_bytes(&self) -> Vec<u8> {
+    let mut res = Vec::new();
```

Comment:
it might be useful to do a `Vec::with_capacity(self.len())` here

Reply:
mmmh, nevermind, since that vector is then used in `HeaderV2::into_bytes()` to fill the vector, the ProxyAddr::into_bytes()` method should take a `&mut Vec<u8>` as argument and the `HeaderV2` method would then do `self.addr.into_bytes(&mut header)`

Reply:
Done